neqo-http3/src/features/extended_connect/connect_udp_session.rs

```diff
@@ -55,32 +55,39 @@ impl Protocol for Session {
         control_stream_recv: &mut Box<dyn RecvStream>,
         now: Instant,
     ) -> Res<Option<State>> {
-        let (f, fin) = self
-            .frame_reader
-            .receive::<ConnectUdpFrame>(
-                &mut StreamReaderRecvStreamWrapper::new(conn, control_stream_recv),
-                now,
-            )
-            .map_err(|_| Error::HttpGeneralProtocolStream)?;
+        loop {
+            let (f, fin) = self
+                .frame_reader
+                .receive::<Capsule>(
+                    &mut StreamReaderRecvStreamWrapper::new(conn, control_stream_recv),
+                    now,
+                )
+                .map_err(|_| Error::HttpGeneralProtocolStream)?;
 
-        if let Some(f) = f {
-            // TODO: Implement HTTP Datagram <https://github.com/mozilla/neqo/issues/2843>.
-            match f {}
-        }
-
-        if fin {
-            events.session_end(
-                ExtendedConnectType::ConnectUdp,
-                self.session_id,
-                CloseReason::Clean {
-                    error: 0,
-                    message: String::new(),
-                },
-                None,
-            );
-            Ok(Some(State::Done))
-        } else {
-            Ok(None)
+            if let Some(f) = f {
```

Comment:
```suggestion
            if let Some(Capsule::Datagram { payload }) = f {
```

Would this compile as well? Saves you one `match` statement.